sql/events.go

```diff
@@ -91,6 +194,301 @@ func (e *EventDetails) CreateEventStatement() string {
+	if e == nil {
+		return "", ""
+	}
+
+	var val, field []string
+	if e.Years != 0 {
+		val = append(val, fmt.Sprintf("%v", e.Years))
+		field = append(field, "YEAR")
+	}
+	if e.Months != 0 {
+		val = append(val, fmt.Sprintf("%v", e.Months))
+		field = append(field, "MONTH")
+	}
+	if e.Days != 0 {
+		val = append(val, fmt.Sprintf("%v", e.Days))
+		field = append(field, "DAY")
+	}
+	if e.Hours != 0 {
+		val = append(val, fmt.Sprintf("%v", e.Hours))
+		field = append(field, "HOUR")
+	}
+	if e.Minutes != 0 {
+		val = append(val, fmt.Sprintf("%v", e.Minutes))
+		field = append(field, "MINUTE")
+	}
+	if e.Seconds != 0 {
+		val = append(val, fmt.Sprintf("%v", e.Seconds))
+		field = append(field, "SECOND")
+	}
+
+	if len(val) == 0 {
+		return "", ""
+	} else if len(val) == 1 {
+		return val[0], field[0]
+	}
+
+	return fmt.Sprintf("'%s'", strings.Join(val, ":")), strings.Join(field, "_")
+}
+
+// EventOnScheduleEveryIntervalFromString returns *EventOnScheduleEveryInterval parsing given interval string
+// such as `2 DAY` or `'1:2' MONTH_DAY`. This function is used in Dolt to construct EventOnScheduleEveryInterval value
+// for the EventDetails.
+func EventOnScheduleEveryIntervalFromString(every string) (*EventOnScheduleEveryInterval, error) {
+	errCannotParseEveryInterval := fmt.Errorf("cannot parse ON SCHEDULE EVERY interval: `%s`", every)
+	strs := strings.Split(every, " ")
+	if len(strs) != 2 {
+		return nil, errCannotParseEveryInterval
+	}
+	intervalVal := strs[0]
+	intervalField := strs[1]
+
+	intervalVal = strings.TrimSuffix(strings.TrimPrefix(intervalVal, "'"), "'")
+	iVals := strings.Split(intervalVal, ":")
+	iFields := strings.Split(intervalField, "_")
+
+	if len(iVals) != len(iFields) {
+		return nil, errCannotParseEveryInterval
+	}
+
+	var interval = &EventOnScheduleEveryInterval{}
+	for i, val := range iVals {
+		n, err := strconv.ParseInt(val, 10, 64)
+		if err != nil {
+			return nil, errCannotParseEveryInterval
+		}
+		switch iFields[i] {
+		case "YEAR":
+			interval.Years = n
+		case "MONTH":
+			interval.Months = n
+		case "DAY":
+			interval.Days = n
+		case "HOUR":
+			interval.Hours = n
+		case "MINUTE":
+			interval.Minutes = n
+		case "SECOND":
+			interval.Seconds = n
+		default:
+			return nil, errCannotParseEveryInterval
+		}
+	}
+
+	return interval, nil
+}
+
+// -------------------------
+//  Events datetime parsing
+// -------------------------
+
+var ErrIncorrectValue = errors.NewKind("Incorrect %s value: '%s'")
+var dateRegex = regexp.MustCompile(`(?m)^(\d{1,4})-(\d{1,2})-(\d{1,2})(.*)$`)
+var timeRegex = regexp.MustCompile(`(?m)^([ T])?(\d{1,2})?(:)?(\d{1,2})?(:)?(\d{1,2})?(\.)?(\d{1,6})?(.*)$`)
+var tzRegex = regexp.MustCompile(`(?m)^([+\-])(\d{2}):(\d{2})$`)
+
+// GetTimeValueFromStringInput returns time.Time in system timezone (SYSTEM = time.Now().Location()).
+// evaluating valid MySQL datetime and timestamp formats.
+func GetTimeValueFromStringInput(field, t string) (time.Time, error) {
```

Comment:
Is there existing code we could use to parse a string timestamp or does this have unique behavior? Could we use `Datetime.Convert(string)` instead? If not, would be good to document why this code needs to be different; it seems like the datetime parsing should be the same across MySQL.

Reply:
If I remember correctly, I added this method because our current method of parsing datetime or timestamp type values does not match MySQL. It is mentioned as TODO https://github.com/dolthub/go-mysql-server/blob/main/sql/types/datetime.go#L232

The reason I wrote incomplete (the timestamp string parsing is missing) parsing method that is different from our current one here is that event queries use datetime value very commonly and that it needs to parse some formats that our current `Datetime.Convert(string)` method is not able to parse. This caused many event query cases to fail.